### PR TITLE
[Core] Support serving encoder/decoder models

### PR DESCRIPTION
**Pull request by DarkLight1337**
This PR cleans up the parsing logic introduced in #4942 to pass `mypy`, and updates the async engine to follow the same code structure, enabling encoder/decoder models to be served using the OpenAI-compatible server. I have added a basic test accordingly.

Other related changes:
- Moved parsing-related code from `vllm.inputs.data` to a new module `vllm.inputs.parse`. To declutter import statements, these functions now have to be imported from `vllm.inputs.parse` explicitly rather than from `vllm.inputs`.
- Most (but not all) of the warnings spammed by `test_bart.py` due to mismatched text have been fixed by padding the vLLM output string with BOS/EOS tokens.
- Added `is_list_of` helper to `vllm.utils` (ported from [#7126](https://github.com/vllm-project/vllm/pull/7126)). This avoids the need for type casts in `parse_and_batch_prompt`.
- Replaced all existing uses of `TypeGuard` with the newly-introduced [`TypeIs`](https://peps.python.org/pep-0742/) construct. The minimum versions of `mypy` and `typing_extensions` have been bumped accordingly.

cc @afeldman-nm

### CODE REVIEW

File: vllm/entrypoints/openai/logits_processors.py [hunk 1]
```diff
@@ -40,9 +40,11 @@ def _get_allowed_token_ids_logits_processor(
...
 
 
-def logit_bias_logits_processor(logit_bias: Dict[str,
-                                                 float], token_ids: List[int],
-                                logits: torch.Tensor) -> torch.Tensor:
+def logit_bias_logits_processor(
+    logit_bias: Dict[int, float],
+    token_ids: List[int],
+    logits: torch.Tensor,
+) -> torch.Tensor:
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/logits_processors.py:47]: This error appeared after updating `mypy` so I had to fix it in this PR.

File: vllm/utils.py [hunk 2]
```diff
@@ -1145,30 +1162,3 @@ def is_embedding_model_config(model_config) -> bool:
...
-             decoder_prompt) in zip(enc_prompt_list, dec_prompt_list)
-    ]
-
-
-def to_enc_dec_tuple_list(
-    enc_dec_prompts: List[ExplicitEncoderDecoderPrompt],
-) -> List[Tuple[PromptInputs, PromptInputs]]:
-    return [(enc_dec_prompt['encoder_prompt'],
-             enc_dec_prompt['decoder_prompt'])
-            for enc_dec_prompt in enc_dec_prompts]
```
Comment (DarkLight1337) [inline vllm/utils.py:1174]: Since `vllm.inputs` could import `utils` but these functions require `vllm.inputs`, I have moved them to `vllm.inputs.parse` to avoid circular imports.

File: vllm/inputs/data.py [hunk 3]
```diff
@@ -1,71 +1,11 @@
...
-                # case 4: array of token arrays
-                return [
-                    ParsedTokens(content=elem, is_tokens=True)
-                    for elem in cast(List[List[int]], prompt)
-                ]
-
-    raise ValueError("prompt must be a string, array of strings, "
-                     "array of tokens, or array of token arrays")
-
-
```
Comment (DarkLight1337) [inline vllm/inputs/data.py:68]: These have been moved to `vllm.inputs.parse`.

File: vllm/inputs/data.py [hunk 4]
```diff
@@ -150,56 +90,6 @@ class ExplicitEncoderDecoderPrompt(TypedDict):
...
-        raise ValueError(f"Invalid prompt {prompt}, valid types are "
-                         "required_keys_dict={required_keys_dict}")
-
-    if isinstance(prompt, str):
-        return "str"
-
-    raise ValueError(f"Invalid prompt {prompt}")
-
-
```
Comment (DarkLight1337) [inline vllm/inputs/data.py:202]: This is no longer required as I have improved the parsing logic inside the engine class itself.

**Review by DarkLight1337 — COMMENTED** [review]
Some explanations

File: vllm/utils.py [hunk 5]
```diff
@@ -1125,50 +1142,3 @@ async def _run_task_with_lock(task: Callable, lock: asyncio.Lock, *args,
...
-                        False)
-
-
-def is_embedding_model_config(model_config) -> bool:
-    '''
-    Extract the embedding model flag from the ModelConfig instance.
-    Return False if model_config is None.
-    '''
-    return model_config is not None and \
-                model_config.embedding_mode
```
Comment (DarkLight1337) [inline vllm/utils.py:1147]: I have moved these two functions to `ModelConfig` for convenience.

File: vllm/inputs/data.py [hunk 6]
```diff
@@ -103,39 +44,41 @@ class TokensPrompt(TypedDict):
...
-    must be SingletonPromptInputs instances.
+    fields of this data structure themselves must be
+    :class:`SingletonPromptInputs` instances.
     """
 
-    encoder_prompt: SingletonPromptInputs
+    encoder_prompt: _T
 
-    decoder_prompt: SingletonPromptInputs
+    decoder_prompt: Optional[_T]
```
Comment (afeldman-nm) [inline vllm/inputs/data.py:81]: Using generics makes sense. Correct me if I am wrong - since the current implementation has a single generic type `_T`, do `encoder_prompt` and `decoder_prompt` have to have the same type?

For example, would this prompt be compatible with `ExplicitEncoderDecoderPrompt`?:

```
{
'encoder_prompt': 'This is an encoder prompt', # raw text prompt
'decoder_prompt': [2] # token list prompt
}
```

The scenario I have in mind is that a vLLM request passes in a raw string encoder prompt (i.e. a question provided by a user) but also wants to explicitly specify a list of control tokens for the decoder prompt in order to affect model behavior.

```suggestion
_T1 = TypeVar("_T1", bound=SingletonPromptInputs, default=SingletonPromptInputs)
_T2 = TypeVar("_T2", bound=SingletonPromptInputs, default=SingletonPromptInputs)


class ExplicitEncoderDecoderPrompt(TypedDict, Generic[_T1,_T2]):
    """Represents an encoder/decoder model input prompt,
    comprising an explicit encoder prompt and a 
    decoder prompt.

    The encoder and decoder prompts, respectively,
    may formatted according to any of the
    :class:`SingletonPromptInputs` schemas, and are not
    required to have the same schema.

    Only the encoder prompt may have multi-modal data.

    Note that an :class:`ExplicitEncoderDecoderPrompt` may not
    be used as an input to a decoder-only model,
    and that the `encoder_prompt` and `decoder_prompt`
    fields of this data structure themselves must be
    :class:`SingletonPromptInputs` instances.
    """

    encoder_prompt: _T1

    decoder_prompt: Optional[_T2]
```

Are there downsides to this approach?

**Review by afeldman-nm — APPROVED** [review]
Hi @DarkLight1337 these improvements are great. I just had a few questions/suggestions. Overall LGTM.

File: vllm/engine/llm_engine.py [hunk 7]
```diff
@@ -792,18 +783,39 @@ def _get_default_enc_dec_decoder_prompt(self, ) -> List[int]:
...
+    def _build_enc_dec_llm_inputs(
+        self,
+        encoder_comps: PromptComponents,
+        decoder_comps: DecoderPromptComponents,
+    ) -> EncoderDecoderLLMInputs:
+        encoder_prompt, encoder_prompt_ids, encoder_mm_data = encoder_comps
+        decoder_prompt, decoder_prompt_ids, decoder_mm_data = decoder_comps
+
+        if encoder_mm_data is not None or decoder_mm_data is not None:
+            raise ValueError("Multi-modal data is not supported for "
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/engine/llm_engine.py:797]: Technically the existing multi-modal models use a vision encoder, so I wanted to make it clear that "encoder" here refers to language encoder.

File: vllm/engine/llm_engine.py [hunk 8]
```diff
@@ -830,136 +842,103 @@ def _process_encoder_decoder_prompt(
...
+                request_id=request_id,
+            )
+
+            if (decoder_input := inputs["decoder_prompt"]) is None:
+                decoder_comps = None, None, None
+            else:
+                decoder_comps = self._extract_prompt_components(
+                    decoder_input,
+                    request_id=request_id,
+                )
```
Comment (afeldman-nm) [inline vllm/engine/llm_engine.py:863]: @DarkLight1337 I see that you already modified `ExplicitEncoderDecoderPrompt` to have two generic types. Still I wanted to answer your question -  this section of `_process_encoder_decoder_prompt()` processes the encoder and decoder prompts independently & does not force them to have the same type.

File: vllm/inputs/data.py (hunk 6 above)
Comment (afeldman-nm) [reply vllm/inputs/data.py]: Hi @DarkLight1337 I see you already made this change, thanks. 

Just to circle back and answer your question - 

This portion of the encoder/decoder example code
https://github.com/vllm-project/vllm/pull/7258/files#diff-9f99d5748f1e063be23804194b63b2b0bb8c1e314d5a35453678e6cbd2996a3aR44-R49

tests a string encoder prompt & an explicit token decoder prompt. It is a supported prompt combination. I also highlighted where in the code this is supported
https://github.com/vllm-project/vllm/pull/7258#discussion_r1707318010

Ultimately when the encoder/decoder prompts reach the backend they are both token lists, so supporting different prompt types is no issue.

File: tests/entrypoints/openai/test_encoder_decoder.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    assert completion.usage == openai.types.CompletionUsage(
+        completion_tokens=5, prompt_tokens=2, total_tokens=7)
+
+    # test using token IDs
+    completion = await client.completions.create(
+        model=model_name,
+        prompt=[0, 0, 0, 0, 0],
+        max_tokens=5,
+        temperature=0.0,
+    )
```
Comment (afeldman-nm) [inline tests/entrypoints/openai/test_encoder_decoder.py:49]: @DarkLight1337 

It is great that we support this now. Does it make sense maybe have two different test prompts, one being a singleton encoder prompt, the other being an explicit encoder/decoder prompt? This would help confirm that the prompt processing pipeline works as it should for the async scenario.

An example prompt could be

```
{
'encoder_prompt': 'The rain in spain',
'decoder_prompt': [0,0,0,0,0]
}
```

Comment (DarkLight1337) [reply tests/entrypoints/openai/test_encoder_decoder.py:49]: I don't think this is specified in the OpenAI API, so users shouldn't run into problems unless they use the async engine directly. Since this PR is pretty big already, let's leave this for future work.

Comment (afeldman-nm) [reply tests/entrypoints/openai/test_encoder_decoder.py:49]: Okay then LGTM.

File: vllm/engine/async_llm_engine.py [hunk 10]
```diff
@@ -334,17 +438,19 @@ async def add_request_async(
         trace_headers: Optional[Mapping[str, str]] = None,
         prompt_adapter_request: Optional[PromptAdapterRequest] = None,
     ) -> None:
+        """Async version of :meth:`add_request`."""
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/async_llm_engine.py:441]: This should not change in this PR

Comment (DarkLight1337) [reply vllm/engine/async_llm_engine.py:441]: Sorry, I don't quite get what you mean by this. Could you elaborate?

**Comment by robertgshaw2-neuralmagic** [comment]
I think this looks good.

Separately @DarkLight1337 (for follow up) perhaps we should create a new class called `AsyncLLMEngineEncoderDecoder` rather than branching inside `AsyncLLMEngine` - WDYT

**Comment by DarkLight1337** [comment]
> I think this looks good.
> 
> Separately @DarkLight1337 (for follow up) perhaps we should create a new class called `AsyncLLMEngineEncoderDecoder` rather than branching inside `AsyncLLMEngine` - WDYT

I think that it would be best to keep the same class structure as the base `LLMEngine`. If the goal is to slim down the engine classes, perhaps we could factor out both the sync and async versions of the parsing code.

File: vllm/engine/async_llm_engine.py [hunk 11]
```diff
@@ -291,38 +295,138 @@ async def stop_remote_worker_execution_loop_async(self) -> None:
...
+            if (decoder_input := inputs["decoder_prompt"]) is None:
+                encoder_comps = await encoder_task
+                decoder_comps = None, None, None
+            else:
+                decoder_task = self._extract_prompt_components_async(
+                    decoder_input,
+                    request_id=request_id,
+                )
+
+                encoder_comps, decoder_comps = await asyncio.gather(
```
Comment (alexm-neuralmagic) [inline vllm/engine/async_llm_engine.py:370]: does the order matter here?

File: vllm/engine/llm_engine.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (afeldman-nm) [reply vllm/engine/llm_engine.py:797]: The point I am about to make it not critical, don't block merging on addressing this - but just for context, the limitation is really that encoder/decoder models *with cross-attention* do not currently support multi-modal.

So for example, a model like Llava has an encoder but not cross-attention.

However, a vision model setup like the one shown in the diagram (taken from a blog post) requires cross attention between ViT and RoBERTa in order to decode a description of the image; even though this is a vision model, it would not currently be supported by vLLM because it requires both multi-modal support *and* encoder/decoder cross-attention, which is a currently-unsupported combination (for example, EncoderDecoderModelRunner has an assert which checks that multimodality is not enabled.)

image

[the remark this answers is not in the document]
Comment (afeldman-nm) [reply vllm/engine/llm_engine.py:797]: Of course this inability to have multimodal+encoder/decoder will need to be addressed by the time we add Whisper, which is a near-term goal

[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/engine/llm_engine.py:797]: Thanks for providing more context, yeah we will need to address this soon.

File: vllm/engine/async_llm_engine.py (hunk 11 above)
Comment (DarkLight1337) [reply vllm/engine/async_llm_engine.py:370]: Yes, that's just how `asyncio.gather` works.

File: vllm/engine/llm_engine.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/engine/llm_engine.py:797]: I'll update the error to say "multi-modal" encoder-decoder models are not supported yet.

File: vllm/engine/async_llm_engine.py (hunk 10 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/engine/async_llm_engine.py:441]: I don't think this function needs changes in this PR - just a nit

Comment (DarkLight1337) [reply vllm/engine/async_llm_engine.py:441]: I see - this is just to keep the order of arguments consistent with the new ordering of parameters in `process_model_inputs_async` (which has been updated alongside `process_model_inputs`).

**Review by robertgshaw2-redhat — APPROVED** [review]

File: vllm/engine/llm_engine.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ywang96) [reply vllm/engine/llm_engine.py]: Just saw this comment - yes like @afeldman-nm said, this is something we should start thinking about since Llama 3.1 multi-modal has exactly the same inference pattern like what Andrew described here.

[the remark this answers is not in the document]
Comment (afeldman-nm) [reply vllm/engine/llm_engine.py]: @ywang96 FYI I just released an RFC giving an overview of next-steps for vLLM's encoder/decoder support, and referenced this discussion in the section on multimodality support: https://github.com/vllm-project/vllm/issues/7366#user-content-considerations-for-designing-multimodal-encoderdecoder-prompt-formats
